src/role-helpers.js

```diff
@@ -94,14 +95,32 @@ function buildElementRoleList(elementRolesMap) {
   }
 
   function getSelectorSpecificity({attributes = []}) {
-    return attributes.length
+    return {
+      primaryKey: attributes.length,
+      secondaryKey: attributes.reduce((acc, {value, constraints = []}) => {
```

Comment:
Could you elaborate about this part a bit more? Ideally some links to how CSS selector specificity is supposed to work by spec.

Reply:
I'm not aware of there being any spec based rules around CSS selector specificity, but happy to look around.

Do you know why there was the choice made for sorting by the number of attributes associated with the role assignment was chosen originally?

---

The issue currently is regarding the `img` element use-cases (and likely other similarly "complex" role assignments). `aria-query` represents each of the three `img` cases with a single attribute object where there is either a constraint for `undefined`, a constraint for `set`, or a `value` of `""`.

The issue arises with the `set` and the `""` cases. Through only sorting by length this lib is inheriting it's specificity ordering for the scenario where attribute lengths are equal somewhat arbitrarily from how the `roles.json` happens to be constructued in `aria-query`.

For a lot things this is fine, but it so happens in this scenario the `set` object is ordered prior to the `""`, and consequently gets evaluated first. The presence of an attribute is a superset of an attribute having a value of `""` so the role `img` is incorrectly applied.

Thinking on it now, the introduction here of a secondary sorting key is potentially a sticking plaster over `makeElementSelector` not being quite on the mark - I might spike seeing if can be more precise there instead.

Reply:
I've reverted and opted for an update to `makeElementSelector` that feels better - see [cab934b](https://github.com/testing-library/dom-testing-library/pull/1241/commits/cab934b9e297d9ae782131009c2d84060c3eeb61).

Reply:
Yeah that reads a lot better, thanks!